components/eventing-controller/controllers/subscription/nats/reconciler.go

```diff
@@ -225,21 +225,21 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	// Check for valid sink
 	if err := r.sinkValidator(ctx, r, subscription); err != nil {
 		log.Errorw("sink URL validation failed", "error", err)
-		if err := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
-			return checkIsConflict(err)
+		if syncErr := r.syncSubscriptionStatus(ctx, subscription, false, false, err.Error()); err != nil {
+			return ctrl.Result{}, syncErr
 		}
-		// No point in reconciling as the sink is invalid
-		return ctrl.Result{}, nil
+		// No point in reconciling as the sink is invalid, return latest error to requeue the reconciliation request
+		return ctrl.Result{}, err
```

Comment:
I did read your comment on the [issue](https://github.com/kyma-project/kyma/issues/13059#issuecomment-1023989101).
Seems this really uses exponential backoff.  
Did you test it with the example from the issue and can confirm it ?

Out of curiosity: Did you dig into the code to the lines where the exponential backoff is implemented?
I came accross [this](https://github.com/kubernetes/client-go/blob/4fc3881b7ffe4a5c1c7c3cffc7dd9d9667989730/util/workqueue/default_rate_limiters.go#L41) but am not sure whether this is really used by the controller.

Reply:
Regarding testing it, I haven't. But I have seen the back-off behaviour work when we return error, since that is a normal flow for our reconciler that sometimes returns errors. Frankly, it is hard to write a test, and even if we do, we are testing controller-runtime.

Yes, that is also where I ended up, checking the behaviour of the default queue.

Reply:
To be honest, IMO, that would be unnecessary. We are testing mostly sink validation logic. The only thing this PR is changing is the return value of the `Reconcile` method in a couple of places. If you really insist on having tests for that, I'd suggest the following:

All we want to verify is that under certain conditions, a single call of `Reconcile` returns error. This can, and should IMO, be done with a simple unit test. We'd need to mock the backend, sink validator and K8s client, and make sure the `Reconcile` call returns error when it should, e.g., when backend has returned error, or sink validator has returned error. The details of the backend and the sink validator are irrelevant in this case, since they are just dependencies of the reconciler.

The problem with using `envtest` in this case is that: 1) we are adding more and more tests for small cases and complicating that test suite more than it is. 2) In the example you explained, we have no control over when or how often is the reconcile method called, and it is affected by the changes to the subscription CRD. E.g. when in step two the subscription status has been updated as not ready, wouldn't that trigger another reconciliation? IMO, `envtest` tests should be reserved for cases when we care about the eventual state of the object (in this case the subscription), not the behaviour/return type of our reconciler code.

With unit test, we could call the Reconcile method exactly one time, they are much faster, and we could test all the cases rather than just one. The problem with the unit test is that there will be a lot of mocking.

Do you agree that this is more suitable for a unit test? Should I try that?

Reply:
As agreed, here is an issue to do this later. Since currently our reconciler tests are 'under construction'!
https://github.com/kyma-project/kyma/issues/13218

Reply:
I talked to @pxsalehi in person and I totally agree a unit test is better than an integration test here.